src/gateway/adhoc.cpp

```diff
@@ -42,66 +44,111 @@ adhoc_t::~adhoc_t() {
+        std::uniform_int_distribution<int> distribution(0, services_by_uuid.size() - 1);
+        std::advance(it, distribution(m_random_generator));
+
+        COCAINE_LOG_DEBUG(m_log, "providing service using remote actor", blackhole::attribute_list({
+            {"uuid", it->second.uuid}
+        }));
+
+        return service_description_t{it->second.endpoints, it->second.protocol, it->second.version};
+    });
 }
 
-size_t
-adhoc_t::consume(const std::string& uuid,
-                 const partition_t& name, const std::vector<asio::ip::tcp::endpoint>& endpoints)
-{
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_exposed(const std::string& uuid,
+                                   const std::string& name,
+                                   unsigned int version,
+                                   const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                   const io::graph_root_t& protocol) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        COCAINE_LOG_DEBUG(m_log, "registering {}/{} destination with {:d} endpoints from {}",
+                          name, version, endpoints.size(), uuid);
+
+        auto& services_by_uuid = remotes[name];
+        bool inserted;
+        std::tie(std::ignore, inserted) = services_by_uuid.insert({uuid, remote_t{uuid, version, endpoints, protocol}});
+
+        if(!inserted) {
+            auto str = format("failed to add remote service {} located on {} to gateway: service already registered", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
+            throw error_t(str);
+        }
 
-    ptr->insert({
-        name,
-        remote_t{uuid, endpoints}
     });
-
-    COCAINE_LOG_DEBUG(m_log, "registering destination with {:d} endpoints", endpoints.size(), attribute_list({
-        {"service", std::get<0>(name)},
-        {"uuid"   , uuid             },
-        {"version", (int)std::get<1>(name)}
-    }));
-
-    return ptr->count(name);
 }
 
-size_t
-adhoc_t::cleanup(const std::string& uuid, const partition_t& name) {
-    remote_map_t::const_iterator lb, ub;
+auto
+adhoc_t::on_local_service_exposed(const std::string& name,
+                                  unsigned int version,
+                                  const std::vector<asio::ip::tcp::endpoint>& endpoints,
+                                  const io::graph_root_t& protocol) -> void {
+    on_remote_service_exposed({}, name, version, endpoints, protocol);
+}
 
-    auto ptr = m_remotes.synchronize();
+auto
+adhoc_t::on_remote_service_removed(const std::string& uuid, const std::string& name) -> void {
+    m_remotes.apply([&](remote_map_t& remotes){
+        if(remotes[name].erase(uuid)) {
+            COCAINE_LOG_DEBUG(m_log, "removed service {} provided by {} from gateway", name, uuid);
+        } else {
+            auto str = format("failed to remove service {} provided by {} from gateway: not found", name, uuid);
+            COCAINE_LOG_ERROR(m_log, str);
```

Comment:
Drop this. We already have an exception which will be eventually caught and logged.